Type traits system

thed636 opened the pull request:
Add type traits system primitives which can be used for writing query parameters and reading binary data.

Commit:
Change apq namespace to libapq

due to possible conflict with internal project

elsid commented on include/apq/query.h:
```diff
@@ -2,7 +2,11 @@
     return boost::hana::make_tuple(std::forward<Args&&>(args) ...);
 }
+
+} // namespace libapq
```
You may just delete this file. It had been commited accidentally.

thed636 replied:
Let it be a placeholder :)

elsid commented on include/apq/apq.h:
```diff
@@ -1,3 +1,7 @@
 
 #include <apq/connection_pool.h>
+
+namespace libapq {
```
Why libapq?

thed636 replied:
Name "apq" conflicts with the internal project name. Do you have any naming suggestions?

elsid replied:
`yapq`?

thed636 replied:
is it really better? :)

dpotepalov replied:
how about pqaxx for namespaces, libpqaxx for the project? Includes a not very subtle "axe"-pun in it... But the real reason I'm suggesting this is because I've googled around for other pq-libraries, and they all seem to be named libpq*xx: libpqxx, libpqmxx, libpq++...

thed636 replied:
We do not provide adaptation of libpq. We provide an advanced convenient mechanism for asynchronous PostgreSQL queries. In this case, "apq" is the best name. But unfortunately, we have a project with this name. I don't like "lib" prefix, to be honest, and I thought about something like apq2, but such name will confuse users. I don't think what to name our project similar to the other projects is a good idea. In this case, a user can't easily distinguish this project from many other ones. It is better to have an original name.

How about "bpq - Better PostgreSQL Queries"? :) Or "acma_pq" with short namespace "acma" (sounds really good!)? Or "amq" - advanced mechanism [for] queries? Or even "pqnyan" - in this case, we can use pretty cat-mascot as a logo? =(^_^)=

elsid commented on tests/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -pedantic -Werror")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -pedantic -Werror -DBOOST_HANA_CONFIG_ENABLE_STRING_UDL=1")
```
Let's do not mix defines and flags in one string literal.

thed636 replied:
Ok

elsid commented on scripts/build_clang_debug.sh:
```diff
@@ -7,7 +7,7 @@ BUILD_DIR=${BUILD_PREFIX}build_clang_debug
 cd ${BUILD_DIR}
 cmake \
-    -DCMAKE_CXX_FLAGS='-std=c++17 -Wall -Wextra -pedantic -Werror' \
+    -DCMAKE_CXX_FLAGS='-std=c++1z -Wall -Wextra -pedantic -Werror' \
```
Why `1z`?

thed636 replied:
Mac OS stock compiler has no such option like C++17. But I can revert this change.

elsid replied:
Too bad for mac. Revert, please.

thed636 replied:
Well, if we want to be user-friendly we must support it - nobody needs a product to compile it only for docker image ;) I'll fix it a little bit later.

thed636 replied:
**Done**

elsid commented on include/apq/type_traits.h:
```diff
@@ -0,0 +1,287 @@
+LIBAPQ_PG_DEFINE_TYPE(std::vector<char>, "bytea", BYTEAOID, dynamic_size)
+
+LIBAPQ_PG_DEFINE_TYPE(boost::uuids::uuid, "uuid", UUIDOID, bytes<16>)
+LIBAPQ_PG_DEFINE_TYPE(std::vector<boost::uuids::uuid>, "[]uuid", 2951, dynamic_size)
```
Why not `uuid[]`?

thed636 replied:
Oops - my mistake. I'll fix it.

Commit:
Add type traits system for C++ to PG mapping

dpotepalov commented on tests/CMakeLists.txt:
```diff
@@ -25,6 +28,7 @@ link_directories("${source_dir}/libs/gherkin-cpp")
 
 set(SOURCES
     connection_pool.cpp
+    type_traits.cpp
```
@thed636 I think you force-pushed the wrong commit as head of the branch, I don't see this file or anything actually related to traits in the PR

thed636 replied:
Oops, my mistake. I've pushed the branch one more time.

dpotepalov commented on include/apq/type_traits.h:
```diff
@@ -0,0 +1,287 @@
+template <typename T>
+struct is_nullable<::boost::scoped_ptr<T>> : ::std::true_type {};
+template <typename T>
+struct is_nullable<::std::unique_ptr<T>> : ::std::true_type {};
```
unique_ptr with custom deleter is also nullable

thed636 replied:
**Done**

dpotepalov commented on include/apq/type_traits.h:
```diff
@@ -0,0 +1,287 @@
+template <typename T>
+inline auto is_null(const T& v) noexcept ->
+    std::enable_if_t<is_nullable<std::decay_t<T>>::value, bool> {
+    return !v;
```
will not work for weak_ptr, specialization which locks prior to checking is required

thed636 replied:
**Done**

dpotepalov commented on include/apq/type_traits.h:
```diff
@@ -0,0 +1,287 @@
+struct is_array : ::std::false_type {};
+
+template <typename T>
+struct is_array<::std::vector<T>> : ::std::true_type {};
```
with custom allocator - also an array

thed636 replied:
**Done**

dpotepalov commented on include/apq/type_traits.h:
```diff
@@ -0,0 +1,287 @@
+template <typename T>
+struct is_array<::std::list<T>> : ::std::true_type {};
+template <typename T>
+struct is_array<::std::set<T>> : ::std::true_type {};
```
I understand what are you trying to accomplish, but this statement feels just plain wrong. With list it's sort of okay, but set is not suitable for accepting an array of values, since an array can contain duplicates, and set cannot. Therefore if the database returns an array and user supplies a set to us, we will essentially lie about the result.

thed636 replied:
Yes, you right. I think we should allow to represent it like an array for serialization but with different type compatibility mechanism.

dpotepalov commented on include/apq/type_traits.h:
```diff
@@ -0,0 +1,287 @@
+
+
+/**
+* Type traits template forvard declaration. Type traits contains information
```
forWard

thed636 replied:
**Done**

dpotepalov commented on include/apq/type_traits.h:
```diff
@@ -0,0 +1,287 @@
+* Type traits template forvard declaration. Type traits contains information
+* related to it's representation in the database. There are two different
+* kind of traits - built-in types with constant OIDs and custom types with
+* database depended OIDs. The functions below describe neccesary traits.
```
dependeNT

thed636 replied:
**Done**

I need a spellchecker for comments :) @elsid can we get some pre-commit grammar-nazi bot? :)

dpotepalov commented on include/apq/type_traits.h:
```diff
@@ -0,0 +1,287 @@
+* Indicates if type is a composite
+*/
+template <typename T>
+struct is_composite : std::integral_constant<bool,
```
Is this trait really necessary? It seems to return true for a lot of types and probably lies about some of them. For example, any user-defined type that is not a composite is composite, judging by this trait.

thed636 replied:
Yes, you partially right, I think the better predicate is to determine if user adapt his type via boost::hana (or maybe boost::fusion as legacy). But anyway, a user always can specialize this template for his (or hers) type.

thed636 replied:
**Done**

thed636 reviewed (commented):
Uploaded new version

dpotepalov committed:
Fixed typo and grammar in comments

dpotepalov reviewed (approved):
I fixed some errors in comments, decided that would be faster than leaving messages in PR. Besides those, looks ok to me.

Commit:
FIX build tests

find_package(PostgreSQL) goes wrong with PostgreSQL_TYPE_INCLUDE_DIR
directory, so the best solution is to define it manually.

elsid commented on include/apq/type_traits.h:
```diff
@@ -0,0 +1,317 @@
+* define this name.
+*/
+template <typename T>
+constexpr auto type_name(const T&) noexcept {
```
We need function that returns `boost::hana::string` to support compile time query text generation.

thed636 replied:
If you are talking about query for custom type oids then I think it is better to use this for every type in the oid map:

    typename type_traits<T>::name{};
